File: DP/longest_increasing_subsequence.cpp
#include<iostream>

using namespace std;

int min(int a, int b);


int main() {
	int n;
	cout << "Enter the number of elements: ";
	cin >> n;
	int a[n];
	cout << "Enter the numbers: " << endl;
	for(int i=0;i<n;i++) {
		cin >> a[i];
	}
	int m[n];
	for(int i=0;i<n;i++) {
		m[i]=1;
	}
	for(int i=0;i<n;i++) {
		for(int j=0;j<i;j++) {
			if(a[i]>a[j]) {
				m[i]=max(m[i],m[j]+1);
			}
		}
	}
	int maxx=0;
	cout << "Array: "<<endl;
	for(int i=0;i<n;i++) {
		if(maxx<m[i]) {
			maxx=m[i];
		}
		cout << m[i] <<" ";
	}
	cout << "LIS: " << maxx << endl;
}

int max(int a, int b) {
	return a>b?a:b;
}

File: src/main/java/in/sg/dsa/dp/LongestIncreasingSubsequence.java
package in.sg.dsa.dp;

import java.util.Arrays;
import java.util.Scanner;

public class LongestIncreasingSubsequence {

  public static void main(String[] args) {
    Scanner scanner = new Scanner(System.in);
    System.out.print("Enter the number of elements: ");
    int n = scanner.nextInt();
    int[] a = new int[n];
    System.out.println("Enter the numbers: ");
    for (int i = 0; i < n; i++) {
      a[i] = scanner.nextInt();
    }

    int[] m = new int[n];
    Arrays.fill(m, 1);

    for (int i = 1; i < n; i++) {
      for (int j = 0; j < i; j++) {
        if (a[i] > a[j]) {
          m[i] = Math.max(m[i], m[j] + 1);
        }
      }
    }

    int max = 0;
    System.out.println("Array: ");
    for (int i = 0; i < n; i++) {
      if (max < m[i]) {
        max = m[i];
      }
      System.out.print(m[i] + " ");
    }
    System.out.println("\nLIS: " + max);
    scanner.close();
  }
}
